nominal/_utils/multipart_downloader.py

```diff
@@ -0,0 +1,347 @@
+# multipart_downloader.py
+from __future__ import annotations
+
+import logging
+import math
+import os
+import pathlib
+import threading
+import time
+from concurrent.futures import Future, ThreadPoolExecutor, as_completed
+from dataclasses import dataclass
+from types import TracebackType
+from typing import Callable, Iterable, Sequence, Tuple, Type
+
+import requests
+from requests.adapters import HTTPAdapter
+from typing_extensions import Self
+from urllib3.util.retry import Retry
+
+logger = logging.getLogger(__name__)
+
+
+class PresignedURLProvider:
+    """Thread-safe presigned URL cache that refreshes on schedule or when invalidated.
+
+    fetch_fn: callable to provide a fresh presigned URL
+    ttl: seconds to reuse the cached URL
+    skew: refresh earlier than TTL by this many seconds
+    """
+
+    def __init__(self, fetch_fn: Callable[[], str], ttl: float = 60.0, skew: float = 5.0) -> None:
+        self._fetch_fn = fetch_fn
+        self._ttl = ttl
+        self._skew = skew
+        self._url: str | None = None
+        self._deadline: float = 0.0
+        self._lock = threading.Lock()
+
+    def get(self, *, force: bool = False) -> str:
+        now = time.monotonic()
+        with self._lock:
+            if force or self._url is None or now >= self._deadline:
+                self._url = self._fetch_fn()
+                self._deadline = now + max(0.0, self._ttl - self._skew)
+                logger.debug("Refreshed presigned url with %s second deadline ('%s')", self._deadline, self._url)
+            return self._url
+
+    def invalidate(self) -> None:
+        with self._lock:
+            logger.info("Invalidated presigned URL ('%s')", self._url)
+            self._deadline = 0.0
+
+
+@dataclass(frozen=True)
+class DownloadItem:
+    """Description of a single file download."""
+
+    provider: PresignedURLProvider
+    destination: pathlib.Path
+    part_size: int = 64 * 1024 * 1024  # 64 MiB
```

Comment:
i dont mind the expression if it's used once, but since it's used a few times lets put it into a constant?